# Making the necessary imports

In [1]:
import configparser as cfp
import pandas as pd
import praw

# Importing reddit credentials from a secure file

In [2]:
def get_reddit_credentials():
    config = cfp.ConfigParser()
    config.read('config.ini')
    return config['reddit']['client_id'], config['reddit']['client_secret'], config['reddit']['user']

# Fetching data from reddit to understand the data structure

In [3]:
client_id, client_secret, user = get_reddit_credentials()
reddit = praw.Reddit(client_id=client_id, client_secret=client_secret, user_agent=user)
number_of_posts = 5001								#Number of top posts we want to fetch from a subreddit
indian_sr = reddit.subreddit('India')

top_posts = indian_sr.top(limit=number_of_posts)	#Fetching top n posts
hot_posts = indian_sr.hot(limit=number_of_posts)	#Fetching the hot posts
new_posts = indian_sr.new(limit=number_of_posts)	#Fetching the newest posts

columns = ['id', 'subreddit', 'title', 'upvotes', 'url', 'original', 'num_comments', 'body', 'created_on', 'flair']

list_posts = []

for post in new_posts:
    list_posts.append([post.id, post.subreddit, post.title, post.score, post.url, post.is_original_content, post.num_comments, post.selftext, post.created, post.link_flair_text])

data = pd.DataFrame(list_posts, columns=columns)
data['creation_date'] = pd.to_datetime(data['created_on'], dayfirst = True, unit ='s')
data.drop(['created_on'], axis=1, inplace=True)

In [4]:
data.head(-5)

id subreddit                                              title  \
0    gychj2     india  RBI grants licence to Bank of China to set up ...   
1    gyccoe     india                      Pls take a look and pls share   
2    gycb88     india  That's one small step for a youtuber, one gian...   
3    gyca8n     india  All religions coming together and taking sensi...   
4    gyc7mg     india  Since there have been lots of Bollywood/India ...   
..      ...       ...                                                ...   
872  gvnwrn     india                      Getting back to the workplace   
873  gvnvri     india  With highest Covid-19 cases, these states push...   
874  gvnqnq     india             Should I buy R15 on loan at this time?   
875  gvndic     india  [RANT] I have been fucking up things for years...   
876  gvg22w     india                         Google what are you up to!   

     upvotes                                                url  original  \
0          1  https://economictimes.indiatimes.com/industry/...     False   
1          0                       https://youtu.be/UzfcphQkOUI     False   
2          1  https://www.reddit.com/r/india/comments/gycb88...     False   
3          1  https://www.hindustantimes.com/india-news/kera...     False   
4          0  https://www.reddit.com/r/india/comments/gyc7mg...     False   
..       ...                                                ...       ...   
872        2  https://www.reddit.com/r/india/comments/gvnwrn...     False   
873       12  https://www.hindustantimes.com/india-news/with...     False   
874        0  https://www.reddit.com/r/india/comments/gvnqnq...     False   
875       35  https://www.reddit.com/r/india/comments/gvndic...     False   
876        0  https://www.reddit.com/r/india/comments/gvg22w...     False   

     num_comments                                               body  \
0               0                                                      
1               0                                                      
2               0  https://ibb.co/p0TMCTb\n\nPeople always fought...   
3               0                                                      
4               1  Average male height in India as per [this](htt...   
..            ...                                                ...   
872             1  As companies come up with innovative communica...   
873             1                                                      
874             7  I very much want to buy the bike, however my s...   
875            29  I am a failure, an utter lazy incompetent fool...   
876             5  ###After Mitron, now Remove Chinese app gets r...   

              flair       creation_date  
0    Policy/Economy 2020-06-07 21:34:37  
1     Non-Political 2020-06-07 21:26:28  
2     Non-Political 2020-06-07 21:23:59  
3       Coronavirus 2020-06-07 21:22:23  
4     Non-Political 2020-06-07 21:17:51  
..              ...                 ...  
872   Non-Political 2020-06-03 13:38:01  
873     Coronavirus 2020-06-03 13:35:53  
874        AskIndia 2020-06-03 13:24:47  
875   Non-Political 2020-06-03 12:57:27  
876   Non-Political 2020-06-03 05:17:17  

[877 rows x 10 columns]

In [5]:
data['flair'].value_counts().sort_values(ascending=False)

Non-Political         217
Politics              200
AskIndia              185
Coronavirus           133
Policy/Economy         50
Business/Finance       30
Science/Technology     21
CAA-NRC-NPR            11
Scheduled              10
Food                    9
Photography             7
Sports                  4
Unverified              3
Megathread              1
Name: flair, dtype: int64

In [6]:
data.shape

(882, 10)

# Collecting More Data

The data collected is only 882 posts. Need a lot more data than that. So, let's pick content on basis of flairs. 

In [7]:
flairs = ["Non-Political", "Politics", "Coronavirus", "Photography", "[R]eddiquette", 
          "Sports", "Food", "Business/Finance", "Policy/Economy", "Science/Technology",
           "AskIndia", "AMA"]
features = ["Title", "ID", "Upvotes", "URL", "Num_comments", "Creation Date", "Body", "Is_original", 
            "Comments", "Flair"]

In [17]:
list_posts = []

for flair in flairs:
    subs = indian_sr.search(f"flair_name:{flair}", limit=300)
    for sub in subs:
        comments = []
        sub.comments.replace_more(limit=10)
        for comment in sub.comments:
            comments.append(str(comment.body))
        list_posts.append([str(sub.title), str(sub.id), sub.score, str(sub.url), sub.num_comments, str(sub.created), 
                             str(sub.selftext), sub.is_original_content, comments, str(sub.link_flair_text)])

data = pd.DataFrame(list_posts, columns = features)

In [23]:
data['Creation Date'] = pd.to_datetime(data['Creation Date'], dayfirst = True, unit ='s')
data.head()

Title      ID  Upvotes  \
0  Dear Indian society, you do not possess the mo...  gx3r7p     8279   
1                           Some things never change  gxjc5c     7205   
2  For an industry that has to paint their actors...  gvw1x2     5788   
3                              The Last Conversation  gvofxe     5130   
4  A fruit seller in Delhi left his crates of man...  goi120     4845   

                                                 URL  Num_comments  \
0  https://www.reddit.com/r/india/comments/gx3r7p...           710   
1                          https://imgur.com/pfqzTiY           391   
2                https://i.redd.it/qhqxocwprp251.jpg           510   
3                https://i.redd.it/n8py56mh1n251.jpg           452   
4                    https://v.redd.it/lp0gqu8h2b051           634   

        Creation Date                                               Body  \
0 2020-06-05 20:36:43  I'm a dark skinned south Indian guy born in ea...   
1 2020-06-06 11:18:58                                                      
2 2020-06-03 23:30:14                                                      
3 2020-06-03 14:19:48                                                      
4 2020-05-22 19:59:10                                                      

   Is_original                                           Comments  \
0        False  [Even though what we Indians experience in our...   
1        False  [Oh God. I'd never understand what they get by...   
2        False  [You're forgetting Priyanka Chopra as Mary Kom...   
3        False  [This is so heart breaking! I can barely contr...   
4        False  [That poor fruit vendor lost his last hope of ...   

           Flair  
0  Non-Political  
1  Non-Political  
2  Non-Political  
3  Non-Political  
4  Non-Political

In [27]:
data[:] = data.sample(frac=1).reset_index(drop=True)
data.head()

Title      ID  Upvotes  \
0  Uber driver scammed me by not picking me up an...  cc6wf1       96   
1              Assembly Elections Results Megathread  a54j07      277   
2  Travelled to Kashmir 2 years back. This is the...  ekd70o       19   
3  Reminder to file your Income tax return - 31st...  ct4z8k       59   
4  After eating 900 mice, cat goes on Haj! RJio j...  fgarrj      214   

                                                 URL  Num_comments  \
0  https://www.reddit.com/r/india/comments/cc6wf1...            59   
1  https://www.reddit.com/r/india/comments/a54j07...           856   
2        https://www.youtube.com/watch?v=6WgnfARwOdA             7   
3  https://www.reddit.com/r/india/comments/ct4z8k...            21   
4  https://www.financialexpress.com/opinion/after...            39   

        Creation Date                                               Body  \
0 2019-07-12 13:53:45  Yesterday it was raining heavily and needed to...   
1 2018-12-11 14:58:33  ---\n# Rajasthan\n\nTotal Seats: 199, Majority...   
2 2020-01-05 21:46:20                                                      
3 2019-08-21 03:51:02  Just filing my tax returns, thought of remindi...   
4 2020-03-10 16:46:22                                                      

   Is_original                                           Comments  \
0        False  [In your case, the driver might have come near...   
1        False  [If elections are not one-sided, it's always g...   
2         True      [Cringe, Good Job brother... It looks good !]   
3        False  [Hey, thanks! Skipped my mind., 16 years of ed...   
4        False  [Jio gives free service for six months\n\nIndi...   

              Flair  
0     [R]eddiquette  
1     [R]eddiquette  
2       Photography  
3     [R]eddiquette  
4  Business/Finance

In [39]:
data['Comments'] = [' '.join(sentence).strip() for sentence in data['Comments']]
data['Comments'].head()

0    In your case, the driver might have come near ...
1    If elections are not one-sided, it's always go...
2           Cringe Good Job brother... It looks good !
3    Hey, thanks! Skipped my mind. 16 years of educ...
4    Jio gives free service for six months\n\nIndia...
Name: Comments, dtype: object

In [40]:
data.to_csv('./data.csv')